lib/Index/IndexDatastore.cpp

```diff
@@ -975,43 +964,63 @@ void UnitMonitor::initialize(IDCode unitCode,
+      markOutOfDate(trigger);
     }
   }
 }
 
 UnitMonitor::~UnitMonitor() {}
 
-std::vector<UnitMonitor::OutOfDateTrigger> UnitMonitor::getOutOfDateTriggers() const {
+std::vector<OutOfDateFileTriggerRef>
+UnitMonitor::getUnorderedOutOfDateTriggers() const {
   sys::ScopedLock L(StateMtx);
-  std::vector<OutOfDateTrigger> triggers;
+  std::vector<OutOfDateFileTriggerRef> triggers;
   for (const auto &entry : OutOfDateTriggers) {
-    triggers.push_back(entry.getValue());
+    triggers.push_back(entry.second);
   }
   return triggers;
 }
 
-void UnitMonitor::checkForOutOfDate(sys::TimePoint<> outOfDateModTime, StringRef filePath, bool synchronous) {
+void UnitMonitor::checkForOutOfDate(sys::TimePoint<> outOfDateModTime,
+                                    StringRef filePath, bool synchronous) {
   sys::ScopedLock L(StateMtx);
   auto findIt = OutOfDateTriggers.find(filePath);
-  if (findIt != OutOfDateTriggers.end() && findIt->getValue().outOfDateModTime >= outOfDateModTime) {
+  if (findIt != OutOfDateTriggers.end() &&
+      findIt->second->getModTime() >= outOfDateModTime) {
     return; // already marked as out-of-date related to this trigger file.
   }
-  if (ModTime < outOfDateModTime)
-    markOutOfDate(outOfDateModTime, DependentFileOutOfDateTriggerHint::create(filePath), synchronous);
+  if (ModTime < outOfDateModTime) {
+    markOutOfDate(OutOfDateFileTrigger::create(filePath, outOfDateModTime),
+                  synchronous);
+  }
 }
 
-void UnitMonitor::markOutOfDate(sys::TimePoint<> outOfDateModTime, OutOfDateTriggerHintRef hint, bool synchronous) {
+void UnitMonitor::markOutOfDate(OutOfDateFileTriggerRef trigger,
+                                bool synchronous) {
   {
+    // Note we have to be careful with the memory management here since the
+    // key for OutOfDateTriggers is a reference into the stored trigger value.
     sys::ScopedLock L(StateMtx);
-    OutOfDateTrigger trigger{ hint, outOfDateModTime};
-    auto &entry = OutOfDateTriggers[trigger.getTriggerFilePath()];
-    if (entry.outOfDateModTime >= outOfDateModTime)
-      return; // already marked as out-of-date related to this trigger file.
-    entry = trigger;
+    auto iter = OutOfDateTriggers.find(trigger->getPathRef());
+    if (iter != OutOfDateTriggers.end()) {
+      // If we have the same or newer mod time for this trigger already stored,
+      // we've seen it before, and have already informed the delegate that the
+      // unit is out of date.
+      if (iter->second->getModTime() >= trigger->getModTime())
+        return;
+
+      // We have a newer mod time for the file, update our trigger, and inform
+      // the delegate that the unit is out of date. Note we need to overwrite
+      // the key as well since it references the path stored in the trigger.
+      *iter = {trigger->getPathRef(), trigger};
+    } else {
+      OutOfDateTriggers.insert({trigger->getPathRef(), trigger});
```

Comment:
You could unconditionally call `insert` and check the `bool` in the return value to know whether it inserted or not. This avoids an unnecessary hash table lookup when the value is missing.

Reply:
Heh, I suggested `try_emplace` but yeah, `insert` is the same here 👍

Reply:
Went with `try_emplace` 👍